playground/$test/hmr-context.js

```diff
@@ -0,0 +1,188 @@
+    }
+  }
+
+  const initServer = async () => {
+    const options = {
+      root: projectTmpDir,
+      logLevel: 'silent',
+      server: {
+        watch: {
+          // During tests we edit the files too fast and sometimes chokidar
+          // misses change events, so enforce polling for consistency
+          usePolling: true,
+          interval: 100,
+        },
+        // host: true,
+        // fs: {
+        //   strict: !isBuild,
+        // },
+      },
+    }
+    const config = await loadConfigFromFile(
+      { command: 'serve', mode: 'development' },
+      undefined,
+      projectTmpDir
+    )
+    const testConfig = mergeConfig(options, config || {})
+
+    const server = await createServer(testConfig)
+
+    await server.listen()
+
+    if (isKeep) {
+      server.printUrls()
+    }
+
+    const baseUrl = server.resolvedUrls.local[0]
+
+    return {
+      server,
+      baseUrl,
+      close: async () => {
+        await server.close()
+      },
+    }
+  }
+
+  const initBrowser = async () => {
+    const browser = await chromium.connect(wsEndpoint, {
+      ...(process.env.OPEN && {
+        slowMo: 100,
+      }),
+    })
+
+    const page = await browser.newPage()
+
+    return {
+      page,
+      close: async () => {
+        await browser?.close()
+      },
+    }
+  }
+
+  const cleanups = []
+
+  /**
+   * @param {{
+   *   files?: Record<string, string>
+   *   exposeFunction?: Record<string, function>
+   * }} [options]
+   */
+  const start = async ({ files, exposeFunction } = {}) => {
+    const { close: closeFixtures } = await initFixtures()
+
+    /**
+     * @param {string} file
+     * @param {string | ((code: string) => string)} contents
+     */
+    const write = async (file, contents) => {
+      const filePath = path.resolve(projectTmpDir, file)
+
+      if (typeof contents === 'function') {
+        const code = await fs.readFile(filePath, 'utf8')
+        contents = contents(code)
+      }
+
+      await fs.writeFile(filePath, contents, 'utf8')
+    }
+
+    if (files) {
+      await Promise.all(
+        Object.entries(files).map(([file, contents]) => write(file, contents))
+      )
+    }
+
+    const { baseUrl, close: closeServer } = await initServer()
+
+    const { page, close: closeBrowser } = await initBrowser()
+
+    if (exposeFunction) {
+      for (const [name, fn] of Object.entries(exposeFunction)) {
+        await page.exposeFunction(name, fn)
+      }
+    }
+
+    await page.goto(baseUrl)
+
+    if (isKeep) {
+      cleanups.push(async () => {
+        await new Promise(() => {})
+      })
+    } else {
+      cleanups.push(closeFixtures, closeServer, closeBrowser)
+    }
+
+    /** @param {string} file */
+    const getFileUrl = (file) => `/${file}`
+
+    /**
+     * @param {string} file
+     * @param {string | ((code: string) => string)} contents
+     */
+    const edit = async (file, contents) => {
+      const updatedPromise = new Promise((resolve) => {
+        // FIXME?
```

Comment:
fix what? that the code below doesn't include the filename in the resolve check anymore? that could lead to false positives if there are multiple edits in parallel

Reply:
Yes, that's it.

As a matter of fact, this doesn't affect currently existing tests and, even though I don't remember, there was a reason I removed it.

My take is go with this version until we actually have the need because of some new tests... Which might very well be never, since focus will now shift on Svelte 5.